php/libraries/Database.class.inc

```diff
@@ -67,27 +67,30 @@ class Database
 
     /**
      * Singleton design pattern implementation - creates the object
-     * and also connects to the database if necessary
-     *
-     * @param string  $database     the database to select
-     * @param string  $username     the username with which to log into
-     *                              the database server
-     * @param string  $password     the password that matches the username
-     * @param string  $host         the name of the database server
-     * @param boolean $trackChanges boolean determining if changes should be
+     * and also connects to the database if necessary.
+     * NOTE There is no return type declaration for this function as it returns
+     * the type `Database` which is not supported by PHPs built in return types.
+     *
+     * @param string|null $database     the database to select
+     * @param string|null $username     the username with which to log into
+     *                                      the database server
+     * @param string|null $password     the password that matches the username
+     * @param string|null $host         the name of the database server
+     * @param boolean     $trackChanges boolean determining if changes should be
      *                              logged to the history table
      *
-     * @return Database
+     * @return object Database
      * @throws DatabaseException
      * @access public
      */
     static function &singleton(
-        $database=null,
-        $username=null,
-        $password=null,
-        $host=null,
-        $trackChanges=true
+        string $database     = null,
+        string $username     = null,
+        string $password     = null,
+        string $host         = null,
```

Comment:
This is going to fail, null isn't a sting.

Reply:
I ran phan locally and it didn't seem to have an issue.... which is weird cause you're right.

I'll change this and the if statement below.

Reply:
That's.. problematic? Maybe it just checks it at the call site and it's never called without them being explicitly passed?

Reply:
Maybe? Not sure how the internals of phan work.